File: src/ExpoGoogleMaps.types.ts
export type Camera = {
  /** The center of the camera */
  center: Coordinate;
  /** How zoomed in the camera is, higher numbers are more zoomed in;
   * @see {@link https://developers.google.com/maps/documentation/ios-sdk/views#zoom}
   */
  zoom: number;
  /** The heading of the camera */
  bearing?: number;
  /** The viewing angle of the camera @see {@link https://developers.google.com/maps/documentation/ios-sdk/views} */
  viewingAngle?: number;
};

export type Polygon = {
  /** The unique key of the polygon */
  key: string;
  /** The coordinates of the polygon */
  coordinates: Coordinate[];
  /** The stroke color of the polygon */
  strokeColor: string;
  /** The fill color of the polygon */
  fillColor?: string;
};

export type Coordinate = {
  latitude: number;
  longitude: number;
};

export type Insets = {
  top?: number;
  left?: number;
  bottom?: number;
  right?: number;
};

export type FitToBoundsOptions = {
  topRight: Coordinate;
  bottomLeft: Coordinate;
  /** The padding around the coordinates. On android, only a single number is allowed, so the max of the insets is chosen */
  insets?: Insets;
};

/**
 * Animation options for camera movements
 */
export type AnimationOptions = {
  /** The duration in seoncds */
  animationDuration?: number;
  /** The animation function to use - ios only
   * @platform ios
   */
  animationFunction?:
    | "easeIn"
    | "easeOut"
    | "easeInEaseOut"
    | "linear"
    | "default";
};

File: android/src/main/java/expo/modules/googlemaps/ExpoGoogleMapsTypes.kt
package expo.modules.googlemaps

import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class Marker : Record {
    @Field
    val coordinate: Coordinate = Coordinate()

    @Field
    val title: String = ""

    @Field
    val key: String = ""
}

class PolygonRecord : Record {
    @Field
    val key = ""

    @Field
    val fillColor: String = ""

    @Field
    val strokeColor: String = ""

    @Field
    val coordinates: Array<Coordinate> = arrayOf()
}

data class Coordinate(@Field val latitude: Double = 0.0, @Field val longitude: Double = 0.0) :
    Record {
    fun toLatLng(): LatLng {
        return LatLng(latitude, longitude)
    }
}

data class Camera(
    @Field val center: Coordinate, @Field val zoom: Double,
    @Field val bearing: Double, @Field val viewingAngle: Double
) : Record {
    fun toGmsCameraPosition(): CameraPosition {
        return CameraPosition.Builder()
            .target(center.toLatLng())
            .zoom(zoom.toFloat())
            .bearing(bearing.toFloat())
            .tilt(viewingAngle.toFloat())
            .build()
    }

    companion object {
        fun fromGmsCameraPosition(cameraPosition: CameraPosition): Camera {
            return Camera(
                Coordinate(cameraPosition.target.latitude, cameraPosition.target.longitude),
                cameraPosition.zoom.toDouble(),
                cameraPosition.bearing.toDouble(),
                cameraPosition.tilt.toDouble()
            )
        }
    }
}

class Insets : Record {
    @Field
    val top: Double = 0.0

    @Field
    val left: Double = 0.0

    @Field
    val bottom: Double = 0.0

    @Field
    val right: Double = 0.0
}

class AnimateOptions : Record {
    @Field
    val animationDuration: Double = 1.0
}